openedx/core/djangoapps/xmodule_django/models.py

```diff
@@ -181,6 +183,19 @@ class UsageKeyField(OpaqueKeyField):
     KEY_CLASS = UsageKey
 
 
+class UsageKeyWithRunField(UsageKeyField):
+    """
+    Subclass of UsageKeyField that automatically fills in
+    missing `run` values, for old Mongo courses.
+    """
+    def to_python(self, value):
+        value = super(UsageKeyWithRunField, self).to_python(value)
+        if value is not None:
+            assert isinstance(value, UsageKey)
```

Comment:
Yes, the `replace` call happens for all courses.  However, the `fill_in_run` method is implemented only in the Old Mongo modulestore.  But, it's a good point that `replace` does get called for all courses at the moment.  It's unclear whether optimizing this would have a clear benefit, but I have found that the low-level `OpaqueKey` operations seem to consume more CPU than one would expect.  I'll go ahead and check for the lack of a `run` before replacing.